hack/generator/pkg/codegen/pipeline_create_storage_types.go

```diff
@@ -0,0 +1,234 @@
+			for _, d := range types {
+				d := d
+
+				if types.IsArmDefinition(&d) {
+					// Skip ARM definitions, we don't need to create storage variants of those
+					continue
+				}
+
+				if _, ok := types.ResolveEnumDefinition(&d); ok {
+					// Skip Enum definitions as we use the base type for storage
+					continue
+				}
+
+				def, err := visitor.VisitDefinition(d, vc)
+				if err != nil {
+					errs = append(errs, err)
+					continue
+				}
+
+				finalDef := def.WithDescription(descriptionForStorageVariant(d))
+				storageTypes[finalDef.Name()] = finalDef
+			}
+
+			if len(errs) > 0 {
+				err := kerrors.NewAggregate(errs)
+				return nil, err
+			}
+
+			types.AddTypes(storageTypes)
+
+			return types, nil
+		},
+	}
+}
+
+// makeStorageTypesVisitor returns a TypeVisitor to do the creation of dedicated storage types
+func makeStorageTypesVisitor(types astmodel.Types) astmodel.TypeVisitor {
+	factory := &StorageTypeFactory{
+		types: types,
+	}
+
+	result := astmodel.MakeTypeVisitor()
+	result.VisitTypeName = factory.visitTypeName
+	result.VisitObjectType = factory.visitObjectType
+	result.VisitArmType = factory.visitArmType
+
+	factory.visitor = result
+	factory.propertyConversions = []propertyConversion{
+		factory.preserveKubernetesResourceStorageProperties,
+		factory.convertPropertiesForStorage,
+	}
+
+	return result
+}
+
+type StorageTypeFactory struct {
+	types               astmodel.Types
+	propertyConversions []propertyConversion
+	visitor             astmodel.TypeVisitor
+}
+
+// A property conversion accepts a property definition and optionally applies a conversion to make
+// the property suitable for use on a storage type. Conversions return nil if they decline to
+// convert, deferring the conversion to another.
+type propertyConversion = func(property *astmodel.PropertyDefinition, ctx StorageTypesVisitorContext) (*astmodel.PropertyDefinition, error)
+
+func (factory *StorageTypeFactory) visitTypeName(_ *astmodel.TypeVisitor, name astmodel.TypeName, ctx interface{}) (astmodel.Type, error) {
+	vc := ctx.(StorageTypesVisitorContext)
+
+	// Resolve the type name to the actual referenced type
+	actualDefinition, actualDefinitionFound := factory.types[name]
+
+	// Check for property specific handling
+	if vc.property != nil && actualDefinitionFound {
+
+		if et, ok := actualDefinition.Type().(*astmodel.EnumType); ok {
+			// Property type refers to an enum, so we use the base type instead
+			return et.BaseType(), nil
+		}
+	}
+
+	// Map the type name into our storage namespace
+	localRef, ok := name.PackageReference.AsLocalPackage()
+	if !ok {
+		return name, nil
+	}
+
+	storageRef := astmodel.MakeStoragePackageReference(localRef)
+	visitedName := astmodel.MakeTypeName(storageRef, name.Name())
+	return visitedName, nil
+}
+
+func (factory *StorageTypeFactory) visitObjectType(
+	_ *astmodel.TypeVisitor,
+	object *astmodel.ObjectType,
+	ctx interface{}) (astmodel.Type, error) {
+	vc := ctx.(StorageTypesVisitorContext)
```

Comment:
minor: I know we like abbreviations in go, but I think `visitorCtx` and `objectCtx` (rather than vc/oc) might help with readability.

Reply:
(Applies across all of the visit methods in this file)

Reply:
I'll go through and give them better names.